### DISCUSSION TITLE
Extend tcType

### DISCUSSION THREAD

**Discussion by CharlieSnowCode** [thunderclient/thunder-client-support#1205] [Q&A]
I am looking to write custom filters in TypeScript.  Since we can only include JavaScript files in a project, I would like to have the JS function call a TS function.  I started by implementing a new function in tc-types.d.ts  When I call it from my custom .js file, I get intellisense, but when I run it I get "Error in Custom Filter: x - x is not a function".
Has anyone been able to add a .ts file to a project and use it from your custom filter?
Many thanks!

**Comment by rangav** [comment]
Hi @CharlieSnowCode, I haven't tried with typescript.

But can you try to compile the typescript file to a js file and attach the js file?

**Reply by CharlieSnowCode** [reply]
I compiled the TS file using 'tsc', now I get an error in the .ts file "duplicate function implementation".  I then included the new .js file in the collection and get the same thing.

**Reply by rangav** [reply]
You need to compile ts to js file, then if you are using multiple files, then you need to bundle to single file.

**Reply by CharlieSnowCode** [reply]
I created a simple way to recreate the issue.
1. Create a TypeScript file (newTS.ts), with an exported function 'greet'
2. Import that function into the JavaScript (custom filter) file that is a part of my collection.
3. Call that new function from a Test
Error is: 'import' and 'export' may appear only with 'sourceType: module' 

So perhaps there is a limitation of the JS file that's included into a ThunderClient project, as the solution appears to be to mark that file as a 'module' and rename the file to have a `.mjs` extension.
**Reference:**
Contents of newTS.ts
`export function greet(name: string) {
    console.log(`Hello, ${name}!`);
  }`
Import function into JS file
`import { greet } from './newTS';`

**Reply by CharlieSnowCode** [reply]
Should I create a feature request for TypeScript support?
Thanks.

**Reply by rangav** [reply]
Please create a new feature request, I will look into it.
